### PR TITLE
feat: support metadata merge

### PR DESCRIPTION
**Pull request by woorui**
# Description

There are three types of metadata in yomo:

1. Metadata from `Authentication.Authenticate()`, This means that the metadata is built in the control stream.
2. Metadata from the data stream.
3. Metadata in DataFrame.

The three types of metadata mentioned above can be merged and route sfn according to thus metadata.

### CODE REVIEW

File: core/context.go [hunk 1]
```diff
@@ -24,8 +23,6 @@ type Context struct {
 	// Frame receives from client.
 	Frame frame.Frame
 
-	// Metadata is be built from handshake.
-	Metadata metadata.Metadata
```
Comment (woorui) [inline core/context.go:28]: Metadata has been moved to the data stream. The data stream now holds metadata that has been merged from both the connection-level and stream-level metadata.

File: core/control_stream.go [hunk 2]
```diff
@@ -42,17 +43,26 @@ type ControlStream interface {
 	CloseWithError(code uint64, errString string) error
 }
 
+// HandshakeFunc is used by server control stream to handle handshake.
+// The returned metadata will be set for the DataStream that is being opened.
+type HandshakeFunc func(*frame.HandshakeFrame) (metadata.Metadata, error)
```
Comment (woorui) [inline core/control_stream.go:48]: handshake return stream-level metadata.

File: core/control_stream.go [hunk 3]
```diff
@@ -42,17 +43,26 @@ type ControlStream interface {
...
+
+// VerifyAuthenticationFunc is used by server control stream to verify authentication.
+type VerifyAuthenticationFunc func(auth.Object) (metadata.Metadata, bool, error)
```
Comment (woorui) [inline core/control_stream.go:51]: verify authentication returns connection-level metadata.

File: core/metadata/metadata.go [hunk 4]
```diff
@@ -1,21 +1,22 @@
...
-import "github.com/yomorun/yomo/core/frame"
-
-// Metadata is used for storing extra info of the application.
+// Metadata is an interface used to store additional information about the application.
+//
+//	There are three types of metadata in yomo:
+//	 1. Metadata from `Authentication.Authenticate()`, This means that the metadata is built in the control stream.
+//	 2. Metadata from the data stream.
+//	 3. Metadata in DataFrame.
 type Metadata interface {
```
Comment (woorui) [inline core/metadata/metadata.go:10]: Update the interface of metadata.

**Review by wujunzhuo — APPROVED** [review]
Cool~

**Pull request merged**